### ISSUE TITLE
Unexpected behavior with a high number of channels Snowpipe Streaming

### ISSUE THREAD

**Issue by adrien-brunelat-sb** [snowflakedb/snowflake-kafka-connector#880]
Hello,

I tried creating a connector v2.2.1 for 32 topics that had each 10 partitions which amounted to 320 Snowpipe Streaming channels across 32 different tables (which is ok according to Snowflake's documentation). The connector kept on opening channels on Snowflake, then closing them and then opening them again without pushing any data. The connector remains in the Running state but never pushes any data.

I tried doing it again with only 4 topics of 10 partitions and it worked fine.

Do you have any idea why it is? It looks like this is coming from the implementation of the connector. Please ask if you need any more information.

**Comment by adrien-brunelat-sb** [comment]
Is there a limit in the number of channels a connector can manage at the same time? I couldn't find any info on the documentation.

**Comment by adrien-brunelat-sb** [comment]
When I check the log I see no warning or errors, but I noticed a lot of messages like this one:

> [Worker-032bc774ad4c60912] [2024-07-12 09:01:05,186] INFO [game-snowflake-2|task-0] [SF_INGEST] Channel is marked as closed, name=GAME_RAW.DATA_INGESTION_DEV.UI_STACK_UPDATED.GAME_EVENTS_UI_STACK_UPDATED_4, channel sequencer=3, row sequencer=0 (net.snowflake.ingest.streaming.internal.SnowflakeStreamingIngestChannelInternal:58)

**Comment by adrien-brunelat-sb** [comment]
This is linked to the support ticket: https://app.snowflake.com/support/case/00801416 at Snowflake.

**Comment by sfc-gh-gjachimko** [comment]
hi @adrien-brunelat-sb 
thanks for reporting that - we had similiar issue in the past, I'll create a documentation with some recommendations how to tackle such issues. however, I'll try to provide you here with some high level guidelines, how to potentially address this issue. Can you share with me if you are using Snowpipe or SnowpipeStreaming?

Here are some basic guidelines, all relate to your connector configuration file:
1. make sure you have this assignment strategy configured:
```"consumer.override.partition.assignment.strategy": "org.apache.kafka.clients.consumer.CooperativeStickyAssignor"```

This will ensure even load distribution of channels for your tasks.
2. define ```"tasks.max"``` property to a number maximum to your CPU's - best if you could reserve one of the CPU's for kafka-connect, so I'd recommend value of Num CPU's - 1
3. Large number of channels per individual instance of connector may result in exceeding underlying kafka timeouts, so I'd suggest also increasing following values:
```
    "consumer.override.rebalance.timeout.ms": "180000",
    "consumer.override.heartbeat.interval.ms":  "5000",
    "consumer.override.session.timeout.ms":    "15000",
    "consumer.override.max.poll.interval.ms": "1200000"
```
These turned out to work well one of the scenarios we've been working with, they give Kafka a chance to cycle through all the partitions assigned to a given task before rebalancing. 

One more thing, not strictly related to connector config, we've noticed that at least some of the kafka-connect configurations have limited memory assigned to the process to 2GB. Not sure if that's your case, but you could also  bump memory for the host process by providing KAFKA_OPTS env variable with -Xmx switch. 

Let me know if these helped. Could you also share information about data flow rate - in terms of messages per second and size of the data?

**Comment by adrien-brunelat-sb** [comment]
Hello, thanks for all the helpful guidelines.

I'm using Snowpipe Streaming, and we have made some progress with Snowflake's Support in parallel. For now we have simply raised the number of CPUs / workers and it works, but I'm definitely going to experiment with your recommendations as well.

This is a test cluster so the flow of data is minimal, the connector loops before sending any data so I had only one message in the cluster.
